addon/utils/_private/steps.ts

```diff
@@ -0,0 +1,43 @@
+import { Node, Schema } from 'prosemirror-model';
+import { Step, StepResult } from 'prosemirror-transform';
+
+//Based on https://discuss.prosemirror.net/t/changing-doc-attrs/784/22
+export class SetDocAttributeStep extends Step {
+  private prevValue: unknown;
+  static ID = 'setDocAttribute';
+  constructor(readonly key: string, readonly value: unknown) {
+    super();
+  }
+  apply(doc: Node): StepResult {
+    const newDoc = doc.copy(doc.content);
+    this.prevValue = newDoc.attrs[this.key];
+    //@ts-expect-error Nodes are normally immutable in prosemirror, this is an exception
+    newDoc.attrs = {
+      ...newDoc.attrs,
+      [this.key]: this.value,
+    };
+    return StepResult.ok(newDoc);
```

Comment:
do you see a way to avoid this? I think (though not yet tested) this will break a pretty fundamental propery
we can now compare docs by reference, so we know that if `doc === otherdoc`, that they will also contain the exact same contents. With this we have a (admittedly rare, but still) case where that is no longer true. 
I don't see any partcular reason we can't just clone the doc node?

Reply:
The `.copy` call should normally clone the doc node already I think?

Reply:
yes, totally. somehow thought it worked differently, shouldve checked the docs :stuck_out_tongue: